[Xtensa] Initial codegen support from IR

andreisfr opened the pull request:
This PR provides implementation of the basic codegen infra such as TargetFrameLowering, MCInstLower,
    AsmPrinter, RegisterInfo, InstructionInfo,  TargetLowering, SelectionDAGISel.

Migrated from https://reviews.llvm.org/D145658

arsenm commented:
Unfortunately you should open a separate PR for each of these changes

s-barannikov reviewed (commented):
Calling convention implementation needs to be in separate PR. It should support passing arguments on the stack, which requires implementing some methods in *FrameLowering, *InstrInfo and *RegisterInfo, as well as supporting load/store/call/return in ISel.

I would also suggest a separate PR for boilerplate classes, i.e. *RegisterInfo, *InstrInfo, *Subtarget, *FrameLowering, *TargetMachine, *AsmPrinter and *DAGToDAGISel. You can check that everything works by launching llc on an empty *.ll file (it should produce a template assembly (-filetype=asm) and object (-filetype=obj) files without crashing).

Supporting ALU operations should then touch only files related to instruction selection. (There may be other _minor_ changes in other files.)

Changes in MCTargetDesc/ look completely unrelated.

andreisfr commented:
> Calling convention implementation needs to be in separate PR. It should support passing arguments on the stack, which requires implementing some methods in *FrameLowering, *InstrInfo and *RegisterInfo, as well as supporting load/store/call/return in ISel.
> 
> I would also suggest a separate PR for boilerplate classes, i.e. *RegisterInfo, *InstrInfo, *Subtarget, *FrameLowering, *TargetMachine, *AsmPrinter and *DAGToDAGISel. You can check that everything works by launching llc on an empty *.ll file (it should produce a template assembly (-filetype=asm) and object (-filetype=obj) files without crashing).
> 
> Supporting ALU operations should then touch only files related to instruction selection. (There may be other _minor_ changes in other files.)
> 
> Changes in MCTargetDesc/ look completely unrelated.

Thanks for the comment. Do I understand correctly that you are proposing to split this patch into three patches (the last one will be with ALU operations)?

s-barannikov commented:
Yes, that's what I'm proposing.

andreisfr committed:
[Xtensa] Initial codegen support from IR.

Initial codegen support for IR tests. Added basic
implementation of the TargetFrameLowering, MCInstLower,
AsmPrinter, RegisterInfo, InstructionInfo,
TargetLowering, SelectionDAGISel.

andreisfr commented:
> Yes, that's what I'm proposing.

I removed the Calling Convention functionality from this commit and will prepare a separate patch with it.

andreisfr commented:
> Unfortunately you should open a separate PR for each of these changes



> Unfortunately you should open a separate PR for each of these changes

Thanks, I removed redundant commits.

MaskRay commented on llvm/lib/Target/Xtensa/XtensaSubtarget.cpp:
```diff
@@ -0,0 +1,46 @@
+
+XtensaSubtarget::XtensaSubtarget(const Triple &TT, const std::string &CPU,
+                                 const std::string &FS, const TargetMachine &TM)
+    : XtensaGenSubtargetInfo(TT, CPU, /*TuneCPU*/ CPU, FS), TargetTriple(TT),
```
Canonical style is `/*TuneCPU=*/CPU`

andreisfr replied:
Thank you very much for comments. Fixed

MaskRay commented:
`git log -- llvm/lib/Target/M68k/` M68k from 2021 sets a good example for new ports. You may follow it:)

MaskRay commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.h:
```diff
@@ -0,0 +1,43 @@
+class XtensaSubtarget;
+class XtensaInstrInfo : public XtensaGenInstrInfo {
+  const XtensaRegisterInfo RI;
+  XtensaSubtarget &STI;
```
Please ensure that each incremental change does not cause new warnings when compiled with a recent Clang.

You may use https://raw.githubusercontent.com/chromium/chromium/main/tools/clang/scripts/update.py  

```
warning: private field 'STI' is not used [-Wunused-private-field]
```

andreisfr replied:
Fixed, I implemented "getSubtarget()" function.

MaskRay commented on llvm/lib/Target/Xtensa/XtensaMCInstLower.h:
```diff
@@ -0,0 +1,43 @@
+class XtensaAsmPrinter;
+
+class LLVM_LIBRARY_VISIBILITY XtensaMCInstLower {
+  MCContext &Ctx;
```
`warning: private field 'Ctx' is not used [-Wunused-private-field]`

andreisfr replied:
Fxied.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -0,0 +1,60 @@
+}
+
+const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+  return NULL;
```
nullptr

andreisfr replied:
Fxied

arsenm commented on llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp:
```diff
@@ -0,0 +1,72 @@
+                                             int SPAdj, unsigned FIOperandNum,
+                                             RegScavenger *RS) const {
+  report_fatal_error("Eliminate frame index not supported yet");
+  return false;
```
dead return

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp:
```diff
@@ -0,0 +1,72 @@
+
+Register XtensaRegisterInfo::getFrameRegister(const MachineFunction &MF) const {
+  const TargetFrameLowering *TFI = MF.getSubtarget().getFrameLowering();
+  return TFI->hasFP(MF) ? (Xtensa::A15) : Xtensa::SP;
```
No () around register name.

Also should do getSubtarget<XTensaSubtarget>()

andreisfr replied:
Fixed

arsenm reviewed (commented):
lgtm with some nits

gerekon commented:
@MaskRay sorry for bothering you. I am from Espressif LLVM team.

> I think the patch series https://reviews.llvm.org/D145658 may be too fine-grained. It might be better to pack more changes in one PR, otherwise there could be a lot of `-Wunused-private-field` and not many testable units.

Looks like this conflicts with suggestions from @s-barannikov https://github.com/llvm/llvm-project/pull/78548#pullrequestreview-1837076213. As far as I understood he proposed to split it into more fine-grained commits.
Or I miss smth?

arsenm commented:
> @MaskRay sorry for bothering you. I am from Espressif LLVM team.
> 
> > I think the patch series https://reviews.llvm.org/D145658 may be too fine-grained. It might be better to pack more changes in one PR, otherwise there could be a lot of `-Wunused-private-field` and not many testable units.
> 
> Looks like this conflicts with suggestions from @s-barannikov [#78548 (review)](https://github.com/llvm/llvm-project/pull/78548#pullrequestreview-1837076213). As far as I understood he proposed to split it into more fine-grained commits. Or I miss smth?

The split is better to review, but the pieces should build warning free and come with the relevant tests. I think this patch is fine with the warnings fixed

s-barannikov commented on llvm/lib/Target/Xtensa/Xtensa.h:
```diff
@@ -0,0 +1,30 @@
+//===- Xtensa.h - Top-level interface for Xtensa representation -*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
```
Current license header template doesn't include this string. I don't know if this is important or not.

andreisfr replied:
Thank you very much for comments. I removed these strings from all new files.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaMCInstLower.h:
```diff
@@ -0,0 +1,43 @@
+class MachineOperand;
+class XtensaAsmPrinter;
+
+class LLVM_LIBRARY_VISIBILITY XtensaMCInstLower {
```
(optional) Is this class going to be large? If not, I'd propose to combine it with XtensaAsmPrinter. See https://github.com/llvm/llvm-project/commit/11e5a0d290bac87f3290547fa2b0aaffc98a798e

andreisfr replied:
Yes, this class not so large, so I combined it with XtensaAsmPrinter.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaMCInstLower.cpp:
```diff
@@ -0,0 +1,62 @@
+
+using namespace llvm;
+
+XtensaMCInstLower::XtensaMCInstLower(MCContext &ctx,
```
(nit) Variable names should start with a capital letter.

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaRegisterInfo.h:
```diff
@@ -0,0 +1,56 @@
+  bool eliminateFrameIndex(MachineBasicBlock::iterator MI, int SPAdj,
+                           unsigned FIOperandNum,
+                           RegScavenger *RS = nullptr) const override;
+  Register getFrameRegister(const MachineFunction &MF) const override;
```
(nit) I'd find it more readable with vertical spaces between method declarations.

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaRegisterInfo.h:
```diff
@@ -0,0 +1,56 @@
+
+} // end namespace llvm
+
+#endif /* LLVM_LIB_TARGET_XTENSA_REGISTERINFO_H */
```
Use C++-style comments.

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaSubtarget.h:
```diff
@@ -0,0 +1,72 @@
+
+class XtensaSubtarget : public XtensaGenSubtargetInfo {
+private:
+  Triple TargetTriple;
```
```suggestion
  const Triple &TargetTriple;
```

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaSubtarget.h:
```diff
@@ -0,0 +1,72 @@
+  XtensaSubtarget &initializeSubtargetDependencies(StringRef CPU, StringRef FS);
+
+public:
+  XtensaSubtarget(const Triple &TT, const std::string &CPU,
+                  const std::string &FS, const TargetMachine &TM);
+
```
const std::string & -> StringRef

TT parameter is redundant as it can be obtained from TM.

andreisfr replied:
Fixed

s-barannikov replied:
Yes, but this is a very minor issue. Most backends pass it anyway.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaTargetMachine.h:
```diff
@@ -36,10 +37,15 @@ class XtensaTargetMachine : public LLVMTargetMachine {
                       std::optional<CodeModel::Model> CM, CodeGenOptLevel OL,
                       bool JIT);
 
+  const XtensaSubtarget *getSubtargetImpl() const { return &Subtarget; }
```
This method should not exist. Subtarget is per-function.

andreisfr replied:
Fixed. I implemented "getSubtargetImpl()" as it is done in M68k backend.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaTargetMachine.h:
```diff
@@ -36,10 +37,15 @@ class XtensaTargetMachine : public LLVMTargetMachine {
   }
+
+protected:
+  XtensaSubtarget Subtarget;
```
See the comment above, there should be no "default" subtarget.

andreisfr replied:
Fxied

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -0,0 +1,41 @@
+  EmitToStreamer(*OutStreamer, LoweredMI);
+}
+
+// Force static initialization.
```
The comment looks irrelevant.

andreisfr replied:
Fixed

s-barannikov reviewed (approved):
LGTM with comments in XtensaTargetMachine.h addressed.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -0,0 +1,44 @@
+
+// hasFP - Return true if the specified function should have a dedicated frame
+// pointer register.  This is true if the function has variable sized allocas or
+// if frame pointer elimination is disabled.
```
Thank you very much for comments. Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -0,0 +1,58 @@
+
+#define DEBUG_TYPE "xtensa-lower"
+
+XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &tm,
+                                           const XtensaSubtarget &STI)
+    : TargetLowering(tm), Subtarget(STI) {
```
```suggestion
XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
                                           const XtensaSubtarget &STI)
    : TargetLowering(TM), Subtarget(STI) {
```

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -0,0 +1,27 @@
+
+using namespace llvm;
+
+XtensaInstrInfo::XtensaInstrInfo(XtensaSubtarget &sti)
+    : XtensaGenInstrInfo(), RI(sti), STI(sti) {}
```
```suggestion
XtensaInstrInfo::XtensaInstrInfo(XtensaSubtarget &STI)
    : XtensaGenInstrInfo(), RI(STI), STI(STI) {}
```

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.h:
```diff
@@ -0,0 +1,45 @@
+  XtensaSubtarget &STI;
+
+public:
+  XtensaInstrInfo(XtensaSubtarget &STI);
```
The parameter should probably be `const`.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaTargetMachine.h:
```diff
@@ -36,10 +37,14 @@ class XtensaTargetMachine : public LLVMTargetMachine {
                       std::optional<CodeModel::Model> CM, CodeGenOptLevel OL,
                       bool JIT);
 
+  const XtensaSubtarget *getSubtargetImpl(const Function &F) const override;
   TargetPassConfig *createPassConfig(PassManagerBase &PM) override;
   TargetLoweringObjectFile *getObjFileLowering() const override {
```
Could also use vertical spacing.

andreisfr replied:
Fxied

s-barannikov reviewed (approved):
LGTM, final style nits

MaskRay commented:
I'll land this patch on your behalf tomorrow, unless someone objects it, assuming that you do not have write access yet!

andreisfr commented:
> I'll land this patch on your behalf tomorrow, unless someone objects it, assuming that you do not have write access yet!

Thank you!

gerekon commented:
[Tracking upstream PR in Espressif repo](https://github.com/espressif/llvm-project/issues/4)
